File: c/laplace1D.c
#include <stdlib.h>
#include <stdio.h>
#include <tgmath.h>





void laplace1D_rhs(long N,const double* u, double* out);
void driver(double alpha,long iter, long N, double a, double b,double* u_out);











void laplace1D_rhs(long N, const double* u, double* out){

	long i;
	long j;
	/*Left BC.*/
	i=0;
	out[i]=-(u[i+1]-2*u[i]);


	/*Right BC.*/
	i=N-3;
	out[i]=-(-2*u[i]+u[i-1]);


	/*Inner nodes.*/
	i=1;
	for(j=i;j<(N-3);j++){
		out[j]=-(u[j+1]-2*u[j]+u[j-1]);
	}

}


void driver(double alpha, long iter, long N, double a, double b, double* u_out){
	long i;
	/*Set up load vector.*/
	double* f = malloc( (N-2)*sizeof(double) );
	for(i=0;i<N-2;i++){
		f[i]=0.0;
	}

	/*Left BC.*/
	i=0;
	f[i]=a;

	/*Right BC.*/
	i=N-3;
	f[i]=b;

	/*Interior solution storage.*/
	double* u=malloc( (N-2)*sizeof(double) );
	/*Output storage for RHS evaluations.*/
	double* out=malloc( (N-2)*sizeof(double) );

	for(i=0;i<N-2;i++){
		u[i]=0.0;
		out[i]=0.0;
	}

	/*Richardson iterations.*/	
	long it=0;
	while(it<iter){
		laplace1D_rhs(N,u,out);
		for(i=0;i<N-2;i++){
			u[i]=u[i]+alpha*(f[i]-out[i]);
		}
	it++;
	}


	/*Populate output vector.*/
	u_out[0]=a;u_out[N-1]=b;
	for(i=1;i<N-1;i++)
		u_out[i]=u[i-1];



	free(f);
	free(out);
	free(u);

}

File: julia/laplace1D.jl
#N - Number of 1D points to use.
#u - Input for RHS evaluation.
#out - Output of RHS evaluation.
function laplace1D_rhs!(N,u,out)
    
    #Left BC
    i=1
    out[i]=-(u[i+1]-2*u[i])

    #Right BC
    i=N-2
    out[i]=-(-2*u[i]+u[i-1])

    #Vectorized inner nodes. 
    i=2;
    out[i:(N-3)]=-(u[(i+1):(N-2)]-2*u[i:(N-3)]+u[(i-1):(N-4)])
end


function driver(alpha,iter,N,a,b,u_out)
    #Set up load vector.
    f=zeros(N-2)

    #Left BC.
    i=1
    f[i]=a

    #Right BC.
    i=N-2
    f[i]=b

    #Initial guess and residual storage.
    u=zeros(N-2)
    out=zeros(N-2)

    it=0
    while it<iter
     laplace1D_rhs!(N,u,out)
     u=u+alpha*(f-out)
     it=it+1
    end

    u_out[1]=a
    u_out[N]=b
    for i = 2 : N-1
        u_out[i]=u[i-1]
    end
end

#For debugging purposes: construct matrix from stencil
function laplace_matrix(N)

    A=zeros(N-2,N-2)
    E=zeros(N-2)
    out=zeros(N-2)

    #Multiply by identity matrix to get stencil matrix.
    for i = 1 : N-2
        E[i]=1.0
        laplace1D_rhs!(N,E,out)
        A[:,i]=out
        E[i]=0.0
    end

    return A


end
